formats/protobuf/src/main/java/io/cloudevents/protobuf/ProtoSerializer.java

```diff
@@ -0,0 +1,214 @@
+
+    /**
+     * Convert the Java SDK CloudEvent into a protobuf representation.
+     *
+     * <p> If the data payload is set with "datacontenttype" set to "application/protobuf", the data provided
+     * must be wrapped in {@link com.google.protobuf.Any}. Storing protobuf data in an Any is required by the
+     * CloudEvent protobuf format spec.
+     *
+     * @param ce The {@link io.cloudevents.CloudEvent} to convert.
+     * @return A {@link CloudEvent} protobuf object that can be serialized to the protobuf binary format.
+     * @throws InvalidProtocolBufferException If the content type of the data is "application/protobuf" but
+     *  the data is not wrapped in {@link com.google.protobuf.Any}.
+     */
+    public static CloudEvent toProto(io.cloudevents.CloudEvent ce) throws InvalidProtocolBufferException {
+        CloudEvent.Builder builder = CloudEvent.newBuilder();
+
+        builder.setSpecVersion(ce.getSpecVersion().toString());
+
+        // Copy the attributes
+        ProtoContextWriter protoContextWriter = new ProtoContextWriter(builder, ce.getSpecVersion());
+        final CloudEventContextReader cloudEventContextReader = CloudEventUtils.toContextReader(ce);
+        cloudEventContextReader.readContext(protoContextWriter);
+
+        // Copy the data
+        final String contentType = ce.getDataContentType();
+        final CloudEventData data = ce.getData();
+        if (data != null) {
+            if (Objects.equals(contentType, PROTO_DATA_CONTENT_TYPE)) {
+                // This will throw if the data provided is not an Any. The protobuf CloudEvent spec requires proto data to be stored as
+                // an Any. I would be amenable to allowing people to also pass raw unwrapped protobufs, but it complicates the logic here.
+                // Perhpas that can be a follow up if there is a need.
+                Any dataAsAny = Any.parseFrom(data.toBytes());
+                builder.setProtoData(dataAsAny);
+            } else if (isTextType(contentType)) {
+                builder.setTextDataBytes(ByteString.copyFrom(data.toBytes()));
+            } else {
+                ByteString byteString = ByteString.copyFrom(data.toBytes());
+                builder.setBinaryData(byteString);
+            }
+        }
+
+        return builder.build();
+    }
+
+    // The proto spec says all text data should go into the text field. It is really difficult to figure out every case
+    // of text data based on the media type though, so I am just going to check for some common cases.
+    private static boolean isTextType(String type) {
+        if (type == null) {
+            return false;
+        }
+        return type.startsWith("text/") || "application/json".equals(type);
+    }
+
+    /**
+     * Defines a {@link CloudEventContextWriter} that will allow setting the attributes within a Protobuf object.
+     */
+    public static class ProtoContextWriter implements CloudEventContextWriter {
```

Comment:
Why not a full `CloudEventWriter`, more than just a `context` writer? It might simplify a bit the above logic I guess

Reply:
I looked into this, and I think it would make the most sense to switch to CloudEventWriter after @JemDay pulls in their `ProtoData` class that implements `CloudEventData`. That way, the `end()` methods can be implemented based on the type of the `data` rather than parsing the datacontenttype attribute.

Reply:
Yep. My vision here is that I want to try, as an overall sdk goal, to have proper `CloudEventWriter`/`CloudEventReader` for every event format/protocol binding, so at some point we might expose these directly: one could just go ahead and pipe an http request into a protobuf `CloudEventWriter`.

But it's a vision, nothing to be done in this PR. If `CloudEventWriter` fit, then you should use it, if it doesn't, that's not a big deal and we can modify it to fit the needs here.